### PR TITLE
sql/catalog: refactor schema handling packages from sql

### PR DESCRIPTION
**Pull request by ajwerner**
This PR attempts to extract the logic related schema management into sub-packages. It should be viewed as a starting point from which we continue to improve. I tried to split the changes into small commits to make it reviewable. Please review this PR as individual commits. 

The layering is such that the `sql/catalog` package is the lowest layer and exports
just interfaces and constants.

Then there are some low-level packages:

   * `sql/catalog/catalogkv` deals with interacting with the relevant system tables
    (other than system.lease), it exposes the `UncachedPhysicalAccessor`.

   * `sql/catalog/database` is a relatively thin wrapper around the `databaseCache`
    which should be going away this release. It also has a dependency on an
    interface exported by schema which is always
    accessors.UncachedPhysicalAccessor

   * `sql/catalog/resolver` deals with name resolution. It depends on the
     accessors through the interfaces defined in `catalog`.

Next we have `sql/catalog/lease` which hosts the `LeaseManager` and `LeaseStore`
and depends on `catalogkv`.

Above that we have `sql/catalog/descs` which really just hosts what was the `sql.TableCollection`. This datastructure is really central in the whole problem. I've renamed it to `descs.Collection`.

Then we have `sql/catalog/accessors` which hosts the `CachedPhysicalAccessor` and the `LogicalAccessor`. This package isn't particularly well abstracted. I suspect we could move the `sql.schemaInterface` here and then unexport the other implementations of `catalog.Accessor`.

Release note: None

### CODE REVIEW

**Comment by ajwerner** [comment]
I think this is RFAL, should be reasonably reviewable. Let me know if you have any issues. Take it one commit at a time.

File: pkg/sql/table.go [hunk 1]
```diff
@@ -258,7 +258,7 @@ func (tc *TableCollection) getMutableTableDescriptor(
 	return mutDesc, nil
 }
 
-// resolveSchemaID attempts to lookup the schema from the schemaCache if it exists,
+// ResolveSchemaID attempts to lookup the schema from the schemaCache if it exists,
```
Comment (knz) [inline pkg/sql/table.go]: the case of the function name in the comment here in 2nd commit does not match the code

File: pkg/sql/catalog/catalogkv/catalogkv.go [hunk 2]
```diff
@@ -0,0 +1,323 @@
+// Copyright 2016 The Cockroach Authors.
```
Comment (knz) [inline pkg/sql/catalog/catalogkv/catalogkv.go:1]: Year in comment (3rd commit)

File: pkg/sql/catalog/accessor.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
+// Copyright 2018 The Cockroach Authors.
```
Comment (knz) [inline pkg/sql/catalog/accessor.go:1]: year in comment (4th commit)

File: pkg/sql/sqltestutils/table_leases.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
+// Copyright 2015 The Cockroach Authors.
```
Comment (knz) [inline pkg/sql/sqltestutils/table_leases.go]: Year in comment (9th commit)

File: pkg/sql/sqltestutils/sql_test_utils.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
+// Copyright 2015 The Cockroach Authors.
```
Comment (knz) [inline pkg/sql/sqltestutils/sql_test_utils.go:1]: year in comment (10th commit)

File: pkg/sql/lease.go [hunk 6]
```diff
@@ -1507,7 +1510,7 @@ func NewLeaseManager(
 	return lm
 }
 
-func nameMatchesTable(
+func NameMatchesTable(
```
Comment (knz) [inline pkg/sql/lease.go:1516]: 12th commit: this function needs a comment

File: pkg/sql/lease.go [hunk 7]
```diff
@@ -2310,6 +2310,44 @@ func (m *LeaseManager) Codec() keys.SQLCodec {
 	return m.codec
 }
 
+// VisitLeases introspects the state of leases managed by the LeaseManager.
+//
+// TODO(ajwerner): consider refactoring the function to take a struct, maybe
+// called LeaseInfo.
+func (m *LeaseManager) VisitLeases(
+	f func(desc sqlbase.TableDescriptor, dropped bool, refCount int, expiration tree.DTimestamp) (wantMore bool),
```
Comment (knz) [inline pkg/sql/lease.go:2319]: :100:  very nice

File: pkg/sql/catalog/lease/helpers_test.go [hunk 8]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2016 The Cockroach Authors.
+// Copyright 2015 The Cockroach Authors.
```
Comment (knz) [inline pkg/sql/catalog/lease/helpers_test.go:1]: year in comment (14th commit)

File: pkg/sql/catalog/lease/main_test.go [hunk 9]
```diff
@@ -0,0 +1,33 @@
+// Copyright 2015 The Cockroach Authors.
```
Comment (knz) [inline pkg/sql/catalog/lease/main_test.go:1]: year in comment (14th commit)

File: pkg/ccl/changefeedccl/schemafeed/schema_feed.go [hunk 10]
```diff
@@ -65,14 +65,14 @@ type Config struct {
 	// change the table at ts1.Next() are emitted as an event.
 	InitialHighWater hlc.Timestamp
 
-	// LeaseManager is used to ensure that when an event is emitted that at a higher
+	// Manager is used to ensure that when an event is emitted that at a higher
```
Comment (knz) [inline pkg/ccl/changefeedccl/schemafeed/schema_feed.go:68]: LeaseManager, not Manager (15th commit "rename types to remove stuttering" )

File: pkg/sql/backfill.go [hunk 11]
```diff
@@ -566,7 +566,7 @@ func (sc *SchemaChanger) validateConstraints(
 	return grp.Wait()
 }
 
-// getTableVersion retrieves the descriptor for the table being
+// GetTableVersion retrieves the descriptor for the table being
```
Comment (knz) [inline pkg/sql/backfill.go:570]: comment mismatch, GetTableVersion is not yet uppercase in commit "export symbols from table collection"

File: pkg/sql/conn_executor_exec.go [hunk 12]
```diff
@@ -661,8 +662,25 @@ func (ex *connExecutor) commitSQLTransactionInternal(
...
+// validatePrimaryKeys verifies that all tables modified in the transaction have
+// an enabled primary key after potentially undergoing DROP PRIMARY KEY, which
+// is required to be followed by ADD PRIMARY KEY.
+func validatePrimaryKeys(tc *TableCollection) error {
+	modifiedTables := tc.GetTablesWithNewVersion()
+	for i := range modifiedTables {
+		table := tc.GetUncommittedTableByID(modifiedTables[i].ID).MutableTableDescriptor
+		if !table.HasPrimaryKey() {
+			return unimplemented.NewWithIssuef(48026,
+				"primary key of table %s dropped without subsequent addition of new primary key",
```
Comment (knz) [inline pkg/sql/conn_executor_exec.go:681]: Is this a new condition? If it is, may be worth isolating in a separate PR beforehand.

File: pkg/sql/logical_schema_accessors.go [hunk 13]
```diff
@@ -87,26 +88,35 @@ func (l *LogicalSchemaAccessor) GetObjectDesc(
 ) (catalog.ObjectDescriptor, error) {
 	switch flags.DesiredObjectKind {
 	case tree.TypeObject:
+		// TODO(ajwerner): Change this function if we ever expose non-table objects
+		// underneath virtual schemas. For now we assume that the only objects
+		// ever handed back from GetObjectByName is
```
Comment (knz) [inline pkg/sql/logical_schema_accessors.go]: sentence is not complete (commit "sql: adopt VirtualSchemas ...")

File: pkg/sql/resolver.go [hunk 14]
```diff
@@ -69,77 +44,94 @@ func (p *planner) ResolveUncachedDatabaseByName(
...
-	requiredType ResolveRequiredType,
-) (res *MutableTableDescriptor, err error) {
-	lookupFlags := tree.ObjectLookupFlags{
-		CommonLookupFlags: tree.CommonLookupFlags{Required: required},
-		RequireMutable:    true,
+type resolveFlags struct {
+	skipCache bool
+}
+
+func (p *planner) ResolveMutableTableDescriptor(
```
Comment (knz) [inline pkg/sql/resolver.go:70]: You may want to keep comment docstrings here and below. (commit "sql/catalog/resolver: extract resolver...")

File: pkg/sql/pg_catalog.go [hunk 15]
```diff
@@ -23,8 +23,8 @@ import (
 
 	"github.com/cockroachdb/cockroach/pkg/keys"
 	"github.com/cockroachdb/cockroach/pkg/security"
+	resolver2 "github.com/cockroachdb/cockroach/pkg/sql/catalog/resolver"
```
Comment (knz) [inline pkg/sql/pg_catalog.go:26]: did you intend to keep "resolver2" here? If yes then maybe give it a better name.

**Review by knz — COMMENTED** [review]
Very nice.

Thanks for splitting the refactor into individual commits. It made reviewing this a breeze, even without Reviewable.

**Comment by knz** [comment]
NB regarding the year in comments:
I appreciate in some cases you just renamed the file and so you may be inclined to keep the existing year. I'll let you decide what you want to do in that case.

But there's at least one or two cases where there were significant changes during the refactor, in that case a new year is warranted.

**Comment by rohany** [comment]
Don't have any real comments -- great work!

**Review by thoszhang — APPROVED** [review]

File: pkg/sql/logical_schema_accessors.go (hunk 13 above)
Comment (ajwerner) [reply pkg/sql/logical_schema_accessors.go]: I fixed up the comment and also the logic, it didn't make sense to construct an `UncachedPhysicalAccessor` here, rather I called through to the `CachedPhysicalAccessor` and had that call through. It should make testing easier when we get around to unit testing this stuff.

File: pkg/sql/resolver.go (hunk 14 above)
Comment (ajwerner) [reply pkg/sql/resolver.go:70]: not sure I follow, what comments did I remove?

File: pkg/sql/conn_executor_exec.go (hunk 12 above)
Comment (ajwerner) [reply pkg/sql/conn_executor_exec.go:681]: Correct, that logic felt like it was in the wrong place and given it had a single call site, didn't deserve to be a method on the `TableCollection`.

https://github.com/cockroachdb/cockroach/blob/d7985c53f1e6d32abd604f5987b9a44f6b210ee3/pkg/sql/table.go#L883-L898

File: pkg/sql/backfill.go (hunk 11 above)
Comment (ajwerner) [reply pkg/sql/backfill.go]: Done.

File: pkg/sql/sqltestutils/sql_test_utils.go (hunk 5 above)
Comment (ajwerner) [reply pkg/sql/sqltestutils/sql_test_utils.go]: Done.

File: pkg/sql/catalog/accessor.go (hunk 3 above)
Comment (ajwerner) [reply pkg/sql/catalog/accessor.go]: Done.

File: pkg/sql/catalog/catalogkv/catalogkv.go (hunk 2 above)
Comment (ajwerner) [reply pkg/sql/catalog/catalogkv/catalogkv.go]: Done.

File: pkg/sql/table.go (hunk 1 above)
Comment (ajwerner) [reply pkg/sql/table.go]: Done.

File: pkg/ccl/changefeedccl/schemafeed/schema_feed.go (hunk 10 above)
Comment (ajwerner) [reply pkg/ccl/changefeedccl/schemafeed/schema_feed.go]: Done.

File: pkg/sql/pg_catalog.go (hunk 15 above)
Comment (ajwerner) [reply pkg/sql/pg_catalog.go]: Done.

File: pkg/sql/catalog/accessor.go [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlbase"
+)
+
+// Accessor provides access to database descriptors.
```
[the remark this answers is not in the document]
Comment (ajwerner) [reply pkg/sql/catalog/accessor.go]: Done.

File: pkg/sql/catalog/catalogkv/physical_accessor.go [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlbase"
+	"github.com/cockroachdb/cockroach/pkg/util/encoding"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+)
+
+// UncachedPhysicalAccessor implements direct access to DB descriptors,
```
[the remark this answers is not in the document]
Comment (ajwerner) [reply pkg/sql/catalog/catalogkv/physical_accessor.go]: Done.

File: pkg/sql/table.go [hunk 18]
```diff
@@ -127,11 +127,11 @@ type TableCollection struct {
 	// table is marked dropped.
 	uncommittedTables []uncommittedTable
 
-	// DatabaseCache is used as a cache for database names.
+	// Cache is used as a cache for database names.
```
[the remark this answers is not in the document]
Comment (ajwerner) [reply pkg/sql/table.go]: Done.

File: pkg/sql/lease.go [hunk 19]
```diff
@@ -519,15 +522,15 @@ func (s LeaseStore) Publish(
 // meant to map to a single immutable descriptor.
 type IDVersion struct {
 	// name only provided for pretty printing.
```
[the remark this answers is not in the document]
Comment (ajwerner) [reply pkg/sql/lease.go]: Done.

File: pkg/sql/zone_test.go [hunk 20]
```diff
@@ -112,7 +112,7 @@ func TestValidSetShowZones(t *testing.T) {
 	sqlutils.VerifyZoneConfigForTarget(t, sqlDB, "DATABASE d", defaultRow)
 	sqlutils.VerifyZoneConfigForTarget(t, sqlDB, "TABLE d.t", defaultRow)
 
-	// Ensure a database zone config applies to that database and its tables, and
+	// Ensure a database zone config applies to that database and its descCollection, and
```
[the remark this answers is not in the document]
Comment (ajwerner) [reply pkg/sql/zone_test.go]: Nope. Fixed.

File: pkg/sql/lease.go (hunk 6 above)
Comment (ajwerner) [reply pkg/sql/lease.go:1516]: Done.

File: pkg/sql/resolver.go (hunk 14 above)
Comment (knz) [reply pkg/sql/resolver.go:70]: The function is exported, so a comment is warranted

File: pkg/sql/logical_schema_accessors.go (hunk 13 above)
Comment (knz) [reply pkg/sql/logical_schema_accessors.go]: Thanks.

**Review by knz — APPROVED** [review]
LGTM

I'll let you resolve the comment thing how you see fit.

**Comment by ajwerner** [comment]
> I'll let you resolve the comment thing how you see fit.

I added some vague comments to those methods. We can make it better later.

**Comment by ajwerner** [comment]
TFTR!

bors r=knz,lucy-zhang

**Comment by ajwerner** [comment]
github outage?

bors r+

**Comment by knz** [comment]
yes looks like it

let's try this again

bors r+
